docs: show overloads for functions

JeanMeche opened the pull request:
Pending angular/dev-infra#2020

Fixes #55556

JoostK commented:
I haven't looked in depth yet, but I'd appreciate if the tooling updates can be pulled out into a dedicated commit, independent of the compiler changes for multiple signatures.

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/entities.ts:
```diff
@@ -134,6 +134,10 @@ export interface FunctionEntry extends DocEntry {
   isNewType: boolean;
 }
 
+export interface FunctionWithOverloadsEntry extends FunctionEntry {
```
We should instead re-use what I've built for initializer APIs: `FunctionWithOverloads`

devversion replied:
I think function extractor should instead return such entry, which has clear `implementation` + `signatures`

JeanMeche replied:
I thought about reusing it too but went with another type because the extractor needs to return a `DocEntry` which `FunctionWithOverloads` doesnt extend.

devversion replied:
I'm thinking about an entry like we have for initializer APIs, with clear separation of implementation + signatures. Maybe we can share that one and use for both? and name it `FunctionEntry`, while the existing `FunctionEntry` becomes `FunctionMetadata` or similar

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -130,24 +130,6 @@ export class DocsExtractor {
-      if (ts.isFunctionDeclaration(declaration)) {
-        const extractor = new FunctionExtractor(exportName, declaration, this.typeChecker);
-        const overloads = extractor
-          .getOverloads()
```
Can we remove this function?

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts:
```diff
@@ -98,3 +118,30 @@ export function extractAllParams(
   }));
 }
+
+export function filterSignatureDeclarations(signatures: readonly ts.Signature[]) {
```
Nit: Re-add JSDoc descriptions. Same below

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts:
```diff
@@ -98,3 +118,30 @@ export function extractAllParams(
+  return result;
+}
+
+export function getOverloads(name: string, typeChecker: ts.TypeChecker, type: ts.Type) {
```
```
extractOverloadSignatures
```

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts:
```diff
@@ -27,24 +27,45 @@ export class FunctionExtractor {
+    const declaration = (symbol?.declarations?.at(-1) as FunctionLike) ?? this.declaration;
+
+    const type = this.typeChecker.getTypeAtLocation(this.declaration);
+    const overloads = filterSignatureDeclarations(type.getCallSignatures()).map((s) => ({
```
Shouldn't this use `getOverloads`/`extractOverloadSignatures`?

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts:
```diff
@@ -27,24 +27,45 @@ export class FunctionExtractor {
+    // In case of overloads, last is the actual implementation with the doc we want to extract
+    const declaration = (symbol?.declarations?.at(-1) as FunctionLike) ?? this.declaration;
+
+    const type = this.typeChecker.getTypeAtLocation(this.declaration);
```
We should potentially rename `this.declaration` to `this.exportDeclaration` for more clarity.

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts:
```diff
@@ -27,24 +27,45 @@ export class FunctionExtractor {
+    const symbol = this.getSymbol();
+
+    // In case of overloads, last is the actual implementation with the doc we want to extract
+    const declaration = (symbol?.declarations?.at(-1) as FunctionLike) ?? this.declaration;
```
I propose calling this `implementation` and use what I did in the initializer API/overload parsing. See `findImplementationOfFunction`. It's safer to _not_ assume the last declaration is the implementation.

JeanMeche committed:
refactor(compiler-cli): extract function overload separatly

in order for the docs to process function entry, this commit refactor function extraction by keeping the implementation as a the default entry and adds all the overloads into a separate array of entries.

JeanMeche commented:
Since this breaks the rendering of the docs, I'll wait for the docs generation to be integrated into this repo to continue.
